kotlin-spark-api/3.2/src/main/kotlin/org/jetbrains/kotlinx/spark/api/jupyter/HtmlRendering.kt

```diff
@@ -0,0 +1,139 @@
+
+        val hasMoreData = tmpRows.size - 1 > numRows
+        val rows = tmpRows.take(numRows)
+
+        tr { th { +"Values" } }
+
+        for (row in rows) tr {
+            td {
+                val string = when (row) {
+                    is ByteArray -> row.joinToString(prefix = "[", postfix = "]") { "%02X".format(it) }
+
+                    is CharArray -> row.iterator().asSequence().toList().toString()
+                    is ShortArray -> row.iterator().asSequence().toList().toString()
+                    is IntArray -> row.iterator().asSequence().toList().toString()
+                    is LongArray -> row.iterator().asSequence().toList().toString()
+                    is FloatArray -> row.iterator().asSequence().toList().toString()
+                    is DoubleArray -> row.iterator().asSequence().toList().toString()
+                    is BooleanArray -> row.iterator().asSequence().toList().toString()
+                    is Array<*> -> row.iterator().asSequence().toList().toString()
+                    is Iterable<*> -> row.iterator().asSequence().toList().toString()
+                    is scala.collection.Iterable<*> -> row.asKotlinIterable().iterator().asSequence().toList().toString()
+                    is Iterator<*> -> row.asSequence().toList().toString()
+                    is scala.collection.Iterator<*> -> row.asKotlinIterator().asSequence().toList().toString()
+                    is Product -> row.productIterator().asKotlinIterator().asSequence().toList().toString()
+                    is Serializable -> row.toString()
+                    // maybe others?
+
+                    is Any? -> row.toString()
+                    else -> row.toString()
+                }
+
+                +string.let {
+                    if (truncate > 0 && it.length > truncate) {
+                        // do not show ellipses for strings shorter than 4 characters.
+                        if (truncate < 4) it.substring(0, truncate)
+                        else it.substring(0, truncate - 3) + "..."
+                    } else {
+                        it
+                    }
+                }
+            }
+        }
+
+        if (hasMoreData) tr {
+            +"only showing top $numRows ${if (numRows == 1) "row" else "rows"}"
+        }
+    }
+} catch (e: SparkException) {
+    // Whenever toString() on the contents doesn't work, since the class might be unknown...
+    """${toString()}
+        |Cannot render this RDD of this class.""".trimMargin()
+}
+
+internal fun <T> Dataset<T>.toHtml(limit: Int = 20, truncate: Int = 30): String = createHtmlTable {
+    val numRows = limit.coerceIn(0 until ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH)
+    val tmpRows = getRows(numRows, truncate).asKotlinList().map { it.asKotlinList() }
+
+    val hasMoreData = tmpRows.size - 1 > numRows
+    val rows = tmpRows.take(numRows + 1)
+
+    tr {
+        for (header in rows.first()) th {
+            +header.let {
```

Comment:
Same here — do we really need this `let`? I would say it's easir to read without it…